sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs

```diff
@@ -406,6 +410,250 @@ protected EventProcessor()
+        ///
+        /// <returns>The <see cref="PartitionProcessor" /> encapsulating the processing task, its cancellation token, and associated state.</returns>
+        ///
+        /// <remarks>
+        ///   This method makes liberal use of class methods and state in addition to the received parameters.
+        /// </remarks>
+        ///
+        internal virtual PartitionProcessor CreatePartitionProcessor(TPartition partition,
+                                                                     EventPosition startingPosition,
+                                                                     CancellationTokenSource cancellationSource)
+        {
+            cancellationSource.Token.ThrowIfCancellationRequested<TaskCanceledException>();
+            var consumer = default(TransportConsumer);
+
+            // If the tracking of the last enqueued event properties was requested, then read the
+            // properties from the active consumer, which can change during processing in the event of
+            // error scenarios.
+
+            LastEnqueuedEventProperties readLastEnquedEventInformation()
+            {
+                if (!Options.TrackLastEnqueuedEventProperties)
+                {
+                    throw new InvalidOperationException(Resources.TrackLastEnqueuedEventPropertiesNotSet);
+                }
+
+                // This is not an expected scenario; the guard exists to prevent a race condition that is
+                // unlikely, but possible, when partition processing is being stopped.
+
+                if (consumer == null)
+                {
+                    throw new InvalidOperationException(Resources.ClientNeededForThisInformationNotAvailable);
+                }
+
+                return new LastEnqueuedEventProperties(consumer.LastReceivedEvent);
+            }
+
+            // Define the routine to handle processing for the partition.
+
+            async Task performProcessing()
+            {
+                var connection = CreateConnection();
+                await using var connectionAwaiter = connection.ConfigureAwait(false);
+
+                var retryPolicy = Options.RetryOptions.ToRetryPolicy();
+                var retryDelay = default(TimeSpan?);
+                var capturedException = default(Exception);
+                var eventBatch = default(IEnumerable<EventData>);
+                var failedAttemptCount = 0;
+                var failedConsumerCount = 0;
+
+                // Continue processing the partition until cancellation is signaled or until the count of failed consumers is too great.
+                // Consumers which been consistently unable to receive and process events will be considered invalid and abandoned for a new consumer.
+
+                while ((!cancellationSource.IsCancellationRequested) && (failedConsumerCount <= MaximumFailedConsumerCount))
```

Comment:
Given you check the cancellation token two lines below, do you need to do it here?

Reply:
There's a double loop going on.    This is the outer loop which manages creating the processor for the first tick and then replacing it with a new instance if the inner loop fails.    The check here says "don't create a consumer and try to process if cancellation has been requested."     The inner loop check says "stop the current processing and bail out" which would normally result in the outer loop believing the consumer is in a failed state and should be recreated.

We could manage the inner by doing a `while(true)` loop and using `cancellationToken.ThrowIfCancelRequested()` but I feel this is a better expression of intent.

Reply:
But CreateConsumer is so fast it that the first check will be just performance overhead. Am I wrong? i.e. is there a semantic difference if the first check (in the outer while) is removed?

Reply:
Quite the opposite;  the `AmqpConsumer` is like `HttpClient`, more or less.   Creating it requires that it use the connection to negotiate opening an AMQP link with the service, and then establish a position in the event stream.   You definitely don't want to treat it as throw-away and do that on each tick.

Because the AMPQ link is persistent and stateful, it can sometimes end up in a bad way after transient failures, service errors, failovers, and similar.   If our processing inner loop hits the point where it feels like it can't make forward progress, we want to try throwing away the consumer and using a fresh one to see if that fixes things.

If we didn't do the outer loop, when the processing isn't making progress, we'd abandon the partition which would put it back in the eligible pool and it would be claimed by a processor again.   Since that can take 5-10 seconds (default values for load balancing), we want to avoid giving up partition ownership if we think that a new consumer will let us keep it.

Reply:
Why is it a sync call them?

Reply:
The create is instantaneous, but the first `ReceiveAsync` call does the heavy lifting to establish the link.    So long as you call `ReciveAsync` on the same instance again, you don't pay the overhead since the link and relevant state is retained between calls.

Reply:
OK, after talking it seems like we could not come up with a simpler way to write this.

Reply:
Chatted with Krzysztof offline and did some brainstorming about ways that we could potentially simplify and improve this flow; we decided that there is sufficient complexity that this approach makes sense in context.